src/box/lua/slab.cc

```diff
@@ -296,5 +426,26 @@ box_lua_slab_init(struct lua_State *L)
 
 	lua_settable(L, -3); /* box.runtime */
 
+	lua_pushstring(L, "malloc");
+	lua_newtable(L);
+
+#if defined(HAVE_MALLOC_INFO)
+	lua_pushstring(L, "info");
+	lua_pushcfunction(L, lbox_malloc_info);
+	lua_settable(L, -3);
+#endif /* defined(HAVE_MALLOC_INFO) */
```

Comment:
Shouldn't we replace it with a stab instead? I would expect that statistics, logging, stringification do its best, but don't raise errors.


I'm not sure here and highlighted to think together.

Reply:
As agreed above, now `box.malloc.info()` always exists and returns `{used = 0, size = 0}` if `malloc_info()` is unavailable or its format is unknown. Full diff below.

<details><summary>diff</summary>

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 669d4760aed5..9b93bb6939af 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -160,11 +160,7 @@ check_function_exists(setproctitle HAVE_SETPROCTITLE)
 check_function_exists(setprogname HAVE_SETPROGNAME)
 check_function_exists(getprogname HAVE_GETPROGNAME)
 
-# If ASAN is enabled, malloc_info() exists but it is not implemented
-# (all counters in the returned document are set to zeros).
-if(NOT ENABLE_ASAN)
-    check_symbol_exists(malloc_info malloc.h HAVE_MALLOC_INFO)
-endif()
+check_symbol_exists(malloc_info malloc.h HAVE_MALLOC_INFO)
 
 #
 # Enable 'make tags' target.
diff --git a/src/box/lua/slab.cc b/src/box/lua/slab.cc
index fc833516f9b4..0006a59f02d1 100644
--- a/src/box/lua/slab.cc
+++ b/src/box/lua/slab.cc
@@ -275,15 +275,18 @@ lbox_slab_check(MAYBE_UNUSED struct lua_State *L)
 	return 0;
 }
 
-#if defined(HAVE_MALLOC_INFO)
 /*
  * Decodes and returns the XML document returned by malloc_info() as is.
+ *
  * This is an internal function that isn't supposed to be called by users,
  * but it may be useful for developers.
+ *
+ * Returns an empty table if malloc_info() isn't supported by the system.
  */
 static int
 lbox_malloc_internal_info(struct lua_State *L)
 {
+#if defined(HAVE_MALLOC_INFO)
 	char *buf;
 	size_t buf_size;
 	FILE *fp = open_memstream(&buf, &buf_size);
@@ -299,6 +302,10 @@ lbox_malloc_internal_info(struct lua_State *L)
 	lua_pushlstring(L, buf, buf_size);
 	free(buf);
 	return luaT_xml_decode(L);
+#else /* !defined(HAVE_MALLOC_INFO) */
+	lua_newtable(L);
+	return 1;
+#endif /* defined(HAVE_MALLOC_INFO) */
 }
 
 /*
@@ -310,10 +317,16 @@ lbox_malloc_internal_info(struct lua_State *L)
  *   }
  *
  * (all numbers are in bytes).
+ *
+ * The information is retrieved with malloc_info(). If it isn't supported by
+ * the system or its format is unknown, {size = 0, used = 0} is returned.
  */
 static int
 lbox_malloc_info(struct lua_State *L)
 {
+	int version = 0;
+	uint64_t total = 0;
+	uint64_t available = 0;
 	lbox_malloc_internal_info(L);
 	/*
 	 * The XML document name returned by malloc_info is expected to be
@@ -321,33 +334,31 @@ lbox_malloc_info(struct lua_State *L)
 	 */
 	assert(lua_istable(L, -1));
 	lua_getfield(L, -1, "malloc");
-	if (lua_istable(L, -1))
-		lua_rawgeti(L, -1, 1);
+	if (!lua_istable(L, -1))
+		goto out;
+	lua_rawgeti(L, -1, 1);
+	if (!lua_istable(L, -1))
+		goto out;
 	/*
 	 * First, check the malloc_info version. The only known version is 1.
 	 * It's pointless to proceed if the version is different.
 	 */
-	int version = 0;
-	if (lua_istable(L, -1)) {
-		lua_getfield(L, -1, "version");
-		version = lua_tonumber(L, -1);
-		lua_pop(L, 1);
-	}
+	lua_getfield(L, -1, "version");
+	version = lua_tonumber(L, -1);
+	lua_pop(L, 1);
 	if (version != 1)
-		return luaL_error(L, "unknown malloc info format");
+		goto out;
 	/*
 	 * Extract the size of used memory. Even though the document version is
 	 * valid, we still need to be careful accessing it.
 	 */
-	uint64_t total = 0;
-	uint64_t available = 0;
 	assert(lua_istable(L, -1));
 	lua_getfield(L, -1, "system");
 	if (lua_istable(L, -1)) {
 		lua_pushnil(L);
 		for (; lua_next(L, -2) != 0; lua_pop(L, 1)) {
 			if (!lua_istable(L, -1))
-				continue;
+				continue; /* skip XML attributes */
 			lua_getfield(L, -1, "type");
 			const char *key = lua_tostring(L, -1);
 			lua_pop(L, 1);
@@ -367,7 +378,7 @@ lbox_malloc_info(struct lua_State *L)
 		lua_pushnil(L);
 		for (; lua_next(L, -2) != 0; lua_pop(L, 1)) {
 			if (!lua_istable(L, -1))
-				continue;
+				continue; /* skip XML attributes */
 			lua_getfield(L, -1, "type");
 			const char *key = lua_tostring(L, -1);
 			lua_pop(L, 1);
@@ -385,6 +396,7 @@ lbox_malloc_info(struct lua_State *L)
 		}
 	}
 	lua_pop(L, 1);
+out:
 	/* Return memory usage information. */
 	lua_newtable(L);
 	luaL_pushuint64(L, total);
@@ -393,7 +405,6 @@ lbox_malloc_info(struct lua_State *L)
 	lua_setfield(L, -2, "used");
 	return 1;
 }
-#endif /* defined(HAVE_MALLOC_INFO) */
 
 /** Initialize box.slab package. */
 void
@@ -429,20 +440,16 @@ box_lua_slab_init(struct lua_State *L)
 	lua_pushstring(L, "malloc");
 	lua_newtable(L);
 
-#if defined(HAVE_MALLOC_INFO)
 	lua_pushstring(L, "info");
 	lua_pushcfunction(L, lbox_malloc_info);
 	lua_settable(L, -3);
-#endif /* defined(HAVE_MALLOC_INFO) */
 
 	lua_pushstring(L, "internal");
 	lua_newtable(L);
 
-#if defined(HAVE_MALLOC_INFO)
 	lua_pushstring(L, "info");
 	lua_pushcfunction(L, lbox_malloc_internal_info);
 	lua_settable(L, -3);
-#endif /* defined(HAVE_MALLOC_INFO) */
 
 	lua_settable(L, -3); /* box.malloc.internal */
 	lua_settable(L, -3); /* box.malloc */
diff --git a/test/app-luatest/gh_7311_malloc_info_test.lua b/test/app-luatest/gh_7311_malloc_info_test.lua
index eea3c0c63ab6..847bc6cca67d 100644
--- a/test/app-luatest/gh_7311_malloc_info_test.lua
+++ b/test/app-luatest/gh_7311_malloc_info_test.lua
@@ -15,8 +15,16 @@ local ALLOC_SIZE = 10 * 1000 * 1000
 local MARGIN = ALLOC_SIZE * 0.05
 
 local function is_supported()
-    return jit.os == 'Linux' and
-           not tarantool.build.flags:match('-fsanitize=[%a,]*address')
+    -- malloc_info() is a GNU extension available only on Linux.
+    if jit.os ~= 'Linux' then
+        return false
+    end
+    -- If ASAN is enabled, malloc_info() exists but it is not implemented
+    -- (all counters in the returned document are set to zeros).
+    if tarantool.build.flags:match('-fsanitize=[%a,]*address') then
+        return false
+    end
+    return true
 end
 
 local function skip_if_supported()
@@ -30,38 +38,27 @@ end
 local function check_malloc_info(info)
     t.assert_ge(info.used, 0)
     t.assert_ge(info.size, info.used)
+    -- It's totally up to the malloc implementation whether to release system
+    -- memory immediately on free() or keep it for future allocations so we can
+    -- only check that the total system memory usage doesn't exceed the actual
+    -- memory usage by more than the size of memory allocated in this test.
+    -- Currently, there are two test cases that allocate ALLOC_SIZE each.
     t.assert_le(info.size, info.used + 2 * ALLOC_SIZE)
 end
 
 g.test_unsupported = function()
     skip_if_supported()
 
-    t.assert_equals(box.malloc, {internal = {}})
+    t.assert_equals(box.malloc.info(), {size = 0, used = 0})
 end
 
 g.test_malloc_info = function()
-    skip_if_unsupported()
-
     t.assert_type(box.malloc, 'table')
     t.assert_type(box.malloc.info, 'function')
+    t.assert_type(box.malloc.info(), 'table')
     t.assert_type(box.malloc.internal, 'table')
     t.assert_type(box.malloc.internal.info, 'function')
-
-    local info = box.malloc.info()
-    t.assert_type(info, 'table')
-    t.assert_type(info.used, 'number')
-    t.assert_type(info.size, 'number')
-    info.used = nil
-    info.size = nil
-    t.assert_equals(info, {})
-
-    info = box.malloc.internal.info()
-    t.assert_type(info, 'table')
-    t.assert_type(info.malloc, 'table')
-    t.assert_type(info.malloc[1], 'table')
-    t.assert_equals(info.malloc[1].version, '1')
-    info.malloc = nil
-    t.assert_equals(info, {})
+    t.assert_type(box.malloc.internal.info(), 'table')
 end
 
 g.test_malloc_small = function()
```

</details>